File: 2024/04-10-ollama-api/demos/ollama-api-test.py
#!/usr/bin/env python

import json
import requests
import sys

def api(payload, url):
    headers = {'Content-Type': 'application/json'}
    json_data = json.dumps(payload)

    response = requests.post(url, data=json_data, headers=headers, stream=True)
    if response.status_code == 200:
        buffer = ""
        try:
            for chunk in response.iter_content(chunk_size=1024):
                # Decode chunk to string
                chunk = chunk.decode('utf-8')
                buffer += chunk
                # Process complete JSON objects
                while '\n' in buffer:
                    line, buffer = buffer.split('\n', 1)  # Split on the first newline
                    if line.strip():
                        try:
                            json_obj = json.loads(line)
                            response_text = json_obj['response']
                            print(response_text, end='', flush=True)  # Print the response part
                        except json.JSONDecodeError:
                            continue  # Ignore JSON decode errors and continue
        except Exception as e:
            print(f"\033[1;31m> Error: {str(e)}\033[0m")
            raise
    else:
        print("\033[1;31m> request failed: {}\033[0m".format(response.status_code))
        raise Exception("Request failed")

if __name__ == "__main__":
    prompt = "Write me a fantasy story!"
    payload = {
        'model': 'mistral',
        'options': {'num_ctx': 16384},
        'prompt': prompt
    }
    api(payload, 'http://localhost:11434/api/generate')

File: 2024/04-10-ollama-api/demos/ollama-api-test.pl
#!/usr/bin/env perl

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ PRAGMAS

use warnings;
use JSON::XS;
use Term::ANSIColor qw|:constants|;
use LWP::UserAgent;
use HTTP::Request;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ INIT

$|++;

my $prompt = "Write me a fantasy story!";

my $payload =
{
	model => 'mistral', 
	options => { num_ctx => 16384 },
	prompt => $prompt,
};

api( $payload, 'http://localhost:11434/api/generate' );

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ SUB

sub api
{
	my $payload = shift;
	my $url = shift;
	my $output;
	
	my $json = JSON::XS->new->utf8;	
	my $ua   = LWP::UserAgent->new;
	
	$ua->timeout(300);
	my $json_data = encode_json( $payload );
	
	my $request = HTTP::Request->new( POST => $url );
	$request->content_type( 'application/json' );
	$request->content( $json_data );
	
	my $response = $ua->request( $request, sub 
	{
		my ($chunk, $res) = @_;
		my $response = $json->decode( $chunk );
		print BOLD GREEN $response->{response}, RESET; 
	} );
	
	unless ( $response->is_success ) 
	{
	    die RED BOLD "> request failed: " . $response->status_line . "\n", RESET;
	}
}

